File: tests/node/backend-contract.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { afterAll, beforeAll, beforeEach, describe, expect, test } from "vitest";
import { REPO_ROOT } from "../support/fixtures.ts";
import { createHarness, type NodeHarness } from "./harness.ts";

/*
 * backend の契約を **Kotlin 実装（server/）と同じ表**で検証する（段階5-1c）。
 *
 * 表は tests/contract/backend-cases.json、散文の正は docs/ARCHITECTURE.md の
 * §4（実測・旧 PHP）と §7（Kotlin の到達点）。
 *
 * ★ ここが無いと、tests/node/harness.ts の仮想 backend は「サーバについての手書きの推測」
 *   でしかない。実際 段階5-1c まで**未知の action に 404 を返していた**（実契約は 501。
 *   php-file の fs 解決に落ちた頃の副産物が残っていた）。同じ表で両側を検証すれば、
 *   仮想 backend は**第 2 実装**になる。
 *
 * ★ 仮想 backend は Map であってファイルシステムではないので、パス解決・dotfile・
 *   拡張子の強制は模せない。**模せる範囲は表の中で `virtual` として宣言してある**ので、
 *   ここでは `virtual: true` だけを流す。残りは Kotlin 側（BackendContractTest）が持つ。
 */

interface ContractRequest {
    readonly method?: string;
    readonly action?: string | null;
    readonly keyword?: string | null;
    readonly body?: string;
    readonly backend?: string;
    readonly trailingSlash?: boolean;
    /** 条件付き更新（段階5-4）。仮想 backend も Kotlin と同じ規則で評価する */
    readonly ifMatch?: string;
    readonly ifNoneMatch?: string;
}

interface ContractCase {
    readonly id: string;
    readonly note?: string;
    readonly seed?: Record<string, string>;
    /** サーバの起動条件（段階5-3 の `readonly` など）。仮想 backend は模さない */
    readonly serverMode?: string;
    readonly request: ContractRequest;
    readonly expect: {
        readonly status: number;
        readonly body?: string;
        readonly headers?: Record<string, string>;
    };
    readonly virtual: boolean;
}

interface ContractTable {
    readonly contractVersion: number;
    readonly cases: readonly ContractCase[];
}

const TABLE: ContractTable = JSON.parse(
    readFileSync(join(REPO_ROOT, "tests", "contract", "backend-cases.json"), "utf8"),
) as ContractTable;

/** 実測 URL の形（docs/ARCHITECTURE.md §4.2）。Kotlin 側の send() と同じ組み立て。 */
function buildUrl(request: ContractRequest): string {
    const backend = request.backend ?? "php-mysql";
    const slash = request.trailingSlash === false ? "" : "/";
    const query: string[] = [];
    if (request.action !== undefined && request.action !== null) {
        query.push("action=" + request.action);
    }
    if (request.keyword !== undefined && request.keyword !== null) {
        query.push("keyword=" + encodeURIComponent(request.keyword));
    }
    return `backend/${backend}${slash}` + (query.length ? "?" + query.join("&") : "");
}

describe("backend の契約（仮想 backend / 段階5-1c）", () => {
    let h: NodeHarness;

    beforeAll(async () => {
        h = await createHarness();
    });

    afterAll(() => {
        h.close();
    });

    beforeEach(() => {
        h.clearServerFiles();
        h.takeRequests();
    });

    /*
     * 仮想 backend に流せるのは「模せる」かつ「通常起動」のケースだけ。
     * `serverMode` を持つもの（段階5-3 の READONLY など）はサーバの起動条件を要求するので、
     * Kotlin 側の専用テスト（ReadOnlyContractTest）が持つ。
     */
    const virtualCases = TABLE.cases.filter((one) => one.virtual && one.serverMode === undefined);

    for (const one of virtualCases) {
        test(`${one.id}`, () => {
            for (const [name, content] of Object.entries(one.seed ?? {})) {
                h.setServerFile(name, content);
            }

            const method = one.request.method ?? "GET";
            const headers: Record<string, string> = {};
            if (one.request.ifMatch !== undefined) {
                headers["If-Match"] = one.request.ifMatch;
            }
            if (one.request.ifNoneMatch !== undefined) {
                headers["If-None-Match"] = one.request.ifNoneMatch;
            }
            const response = h.callBackend(buildUrl(one.request), {
                method: method,
                data: method === "POST" ? (one.request.body ?? "") : undefined,
                headers: headers,
            });

            expect(response.status, `${one.id}: status（${one.note ?? ""}）`).toBe(one.expect.status);

            if (one.expect.body !== undefined) {
                /* 空 body は仮想 backend も実サーバも空文字（XHR の responseText と同じ形） */
                expect(response.data ?? "", `${one.id}: body`).toBe(one.expect.body);
            }

            for (const [name, value] of Object.entries(one.expect.headers ?? {})) {
                expect(response.headers[name], `${one.id}: ヘッダ ${name}`).toBe(value);
            }
        });
    }

    test("表そのものが健全（版・件数・id の重複なし）", () => {
        expect(TABLE.contractVersion).toBe(1);
        expect(TABLE.cases.length).toBeGreaterThan(15);
        const ids = TABLE.cases.map((one) => one.id);
        expect(new Set(ids).size).toBe(ids.length);
    });

    test("模せない範囲が表の中で宣言されている", () => {
        /*
         * `virtual: false` が 0 件になったら、それは「仮想 backend が実サーバと同等になった」
         * のではなく「宣言を書き忘れた」ことのほうがずっと起こりやすい。両方が 1 件以上
         * あることを確かめておく（表の意味が空洞化するのを防ぐ）。
         */
        expect(virtualCases.length).toBeGreaterThan(0);
        expect(TABLE.cases.filter((one) => !one.virtual).length).toBeGreaterThan(0);
    });

    /*
     * ★ check() に**通さない**と決めた status。
     *
     * - 412（If-Match / If-None-Match の不一致。段階5-4）
     *   「衝突したので上書きするか？」は**エラー表示ではなく分岐**。フロントが握って
     *   confirm に流すので、textarea に文言を出してはいけない。プリフライトの 404 を
     *   check() に通さないのと同じ理屈（js/io.ts の preflightresponse）。
     *
     * ここに足すのは「フロントが自分で処理する」status だけ。**表示すべきものを足して
     * 検査から逃がさないこと** —— それをやると、このテストが守っているものが無くなる。
     */
    const HANDLED_BY_FRONTEND = new Set([412]);

    /** 契約表に出てくる異常系の status のうち、check() が知っているべきもの。 */
    function statusesCheckMustKnow(): number[] {
        return Array.from(
            new Set(
                TABLE.cases
                    .map((one) => one.expect.status)
                    .filter((status) => status >= 400)
                    .filter((status) => !HANDLED_BY_FRONTEND.has(status)),
            ),
        ).sort((a, b) => a - b);
    }

    test("表に出てくる異常系の status を js/io.ts の check() が全部知っている", () => {
        /*
         * ★ これが本ファイルを置いた 2 つ目の理由。
         *
         * js/io.ts の check() は「表示すべき応答」を switch で列挙しており、**知らない status は
         * default: return true に落ちて「成功」に倒れる**。§5 で新設した 400 / 403 / 405 を
         * 足し忘れると、ユーザーには何も出ないまま save が失敗する。
         *
         * 200 / 201 は成功側だが、201 は locale の http201 = "Saved" を出す契約なので
         * check() が知っている必要がある（段階4-3b から）。ここでは異常系（>= 400）だけを見る。
         */
        const statuses = statusesCheckMustKnow();

        expect(statuses.length).toBeGreaterThan(0);
        for (const status of statuses) {
            expect(h.io.check(status), `check() が ${status} を知らない`).toBe(false);
        }
    });

    test("フロントが自分で処理する status は check() に通さない", () => {
        /*
         * 412 を check() に足してしまうと、衝突のたびに textarea へ文言が出て
         * confirm の分岐と二重になる。**通さないことも契約**なので固定する。
         */
        for (const status of HANDLED_BY_FRONTEND) {
            expect(h.io.check(status), `check() が ${status} を拾ってしまっている`).toBe(true);
        }
    });

    test("check() が知っている status には locale の文言がある", () => {
        /*
         * check() に case を足しても locale のキーが無ければ、textarea には翻訳されない
         * "http400" という文字列がそのまま出る（_() は未知キーをキー名のまま返す）。
         */
        const en = readFileSync(join(REPO_ROOT, "locale", "en.xml"), "utf8");

        for (const status of statusesCheckMustKnow()) {
            expect(en, `locale/en.xml に http${status} が無い`).toContain(`name="http${status}"`);
        }
    });
});

File: server/src/test/kotlin/io/propagandist/grabado/api/BackendContractTest.kt
package io.propagandist.grabado.api

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import tools.jackson.databind.JsonNode
import tools.jackson.databind.json.JsonMapper
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.ExperimentalPathApi
import kotlin.io.path.deleteRecursively

/**
 * `tests/contract/backend-cases.json` を 1 ケース 1 テストとして流す。
 *
 * **契約を 2 言語で二重に書かないための仕掛け。** 同じ表を段階5-1c で `tests/node/` の
 * 仮想 backend にも流し、harness が「サーバについての手書きの推測」から「同じ表で検証された
 * 第 2 実装」になる。散文の正は `docs/ARCHITECTURE.md` §4（実測）/ §7（到達点）。
 *
 * ## MockMvc ではなく実サーバを使う
 *
 * 契約には**日本語の keyword が URL エンコードで往復すること**と `%2F` の扱いが含まれる。
 * どちらもサーブレットコンテナの URL デコード層の話で、**MockMvc はその層を素通りする**。
 * `RANDOM_PORT` ＋ JDK 標準の [HttpClient] なら実際に通り、しかも依存が 1 つも増えない
 * （Boot 4 では `@AutoConfigureMockMvc` が `spring-boot-starter-test` の外に出ている）。
 *
 * 「1 リクエスト 1 レスポンス」で表せないもの —— 往復のバイト一致、トラバーサルで
 * **外にファイルができないこと**、原子的置換 —— は [BackendBehaviourTest] と
 * `FileDesignStoreTest` が直接書く。
 */
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class BackendContractTest {

    @Value("\${local.server.port}")
    private var port: Int = 0

    @BeforeEach
    fun clean() = clearSchemaDir()

    @ParameterizedTest(name = "{0}")
    @MethodSource("cases")
    fun contract(id: String, case: JsonNode) {
        seed(case)

        val response = send(case.path("request"), port)
        val expect = case.path("expect")
        val note = case.path("note").asString("")

        assertThat(response.statusCode())
            .describedAs("%s: status（%s）", id, note)
            .isEqualTo(expect.path("status").asInt())

        if (expect.has("body")) {
            assertThat(String(response.body(), StandardCharsets.UTF_8))
                .describedAs("%s: body", id)
                .isEqualTo(expect.path("body").asString())
        }

        expect.path("headers").properties().forEach { (name, value) ->
            assertThat(response.headers().firstValue(name).orElse(null))
                .describedAs("%s: ヘッダ %s", id, name)
                .isEqualTo(value.asString())
        }
    }

    /** 表そのものの健全性。ケースが 0 件のまま緑になる事故を防ぐ。 */
    @Test
    fun `契約表は空でなく、id が重複していない`() {
        // JsonNode 自身が map(...) を持つので、Kotlin の Iterable.map は使わない（for で回す）。
        val ids = buildList { for (case in table().path("cases")) add(case.path("id").asString()) }
        assertThat(ids).hasSizeGreaterThan(15)
        assertThat(ids).doesNotHaveDuplicates()
        assertThat(table().path("contractVersion").asInt()).isEqualTo(1)
    }

    private fun seed(case: JsonNode) {
        case.path("seed").properties().forEach { (name, content) ->
            Files.write(schemaDir.resolve(name), content.asString().toByteArray(StandardCharsets.UTF_8))
        }
    }

    companion object {
        /**
         * 正本ディレクトリ。`@TempDir` ではなく自前で作るのは、[DynamicPropertySource] が
         * コンテキスト生成前（＝ `@TempDir` の注入より前）に読まれるため。
         */
        @JvmStatic
        val schemaDir: Path = Files.createTempDirectory("grabado-contract-")

        private val http: HttpClient = HttpClient.newHttpClient()

        @JvmStatic
        @DynamicPropertySource
        fun properties(registry: DynamicPropertyRegistry) {
            registry.add("grabado.schema-dir") { schemaDir.toString() }
        }

        @JvmStatic
        fun clearSchemaDir() {
            Files.newDirectoryStream(schemaDir).use { stream -> stream.forEach(Files::delete) }
        }

        @JvmStatic
        @AfterAll
        @OptIn(ExperimentalPathApi::class)
        fun cleanup(): Unit = schemaDir.deleteRecursively()

        /** 契約表。repo ルートは Gradle が system property で渡す（build.gradle.kts）。 */
        @JvmStatic
        fun table(): JsonNode {
            val repoRoot = Path.of(
                System.getProperty("grabado.repoRoot")
                    ?: error("grabado.repoRoot が未設定（build.gradle.kts の test タスクを見ること）"),
            )
            val file = repoRoot.resolve("tests").resolve("contract").resolve("backend-cases.json")
            check(Files.isRegularFile(file)) { "契約表が無い: $file" }
            return Files.newInputStream(file).use { JsonMapper().readTree(it) }
        }

        /**
         * `@MethodSource` は Iterable も受けるので Stream に変換しない。
         *
         * ★ `JsonNode` は自前の `map(...)` を持つため、Kotlin の `Iterable.map` は解決されない。
         *   `for` で回すのが確実（同じ罠が [contract] の ids 側にもある）。
         */
        @JvmStatic
        fun cases(): List<Arguments> = casesFor(null)

        /**
         * 指定した `serverMode` のケースだけを返す（`null` は「モード指定なし」＝通常起動）。
         *
         * `serverMode` を持つケースは別の起動条件を要求するので、通常のテストでは流さない
         * —— `readonly` は [ReadOnlyContractTest] が持つ。
         */
        fun casesFor(serverMode: String?): List<Arguments> = buildList {
            for (case in table().path("cases")) {
                val mode = if (case.has("serverMode")) case.path("serverMode").asString() else null
                if (mode == serverMode) {
                    add(Arguments.of(case.path("id").asString(), case))
                }
            }
        }

        /**
         * 契約表の `request` を実 HTTP に変換して投げる。
         *
         * URL の形は実測どおり `<xhrpath>backend/<backend名>/?action=...`
         * （`ARCHITECTURE.md` §4.2）。`keyword` は `encodeURIComponent` 相当でエンコードする。
         *
         * ★ **`path` があればそれを使う**（段階11-2a）。`/api/ai/review` は
         *   `/backend/<name>/?action=` の形を取らない —— `/api/` は §11 が始める名前空間で、
         *   upstream の語彙に乗せない（5-0 の決定）。表の側で URL の組み立て方が 2 通りに
         *   なるが、**どちらなのかはデータで分かる**（`request.path` の有無）。
         */
        fun send(request: JsonNode, port: Int): HttpResponse<ByteArray> {
            val backend = request.path("backend").asString("file")
            val slash = if (request.path("trailingSlash").asBoolean(true)) "/" else ""
            val query = buildList {
                if (request.has("action") && !request.path("action").isNull) {
                    add("action=" + request.path("action").asString())
                }
                if (request.has("keyword") && !request.path("keyword").isNull) {
                    add("keyword=" + URLEncoder.encode(request.path("keyword").asString(), StandardCharsets.UTF_8))
                }
            }.joinToString("&")

            val uri = if (request.has("path")) {
                URI.create("http://127.0.0.1:$port" + request.path("path").asString())
            } else {
                URI.create(
                    "http://127.0.0.1:$port/backend/$backend$slash" + if (query.isEmpty()) "" else "?$query",
                )
            }
            val builder = HttpRequest.newBuilder(uri)
            when (val method = request.path("method").asString("GET")) {
                "GET" -> builder.GET()
                "POST" -> builder.POST(
                    HttpRequest.BodyPublishers.ofByteArray(
                        request.path("body").asString("").toByteArray(StandardCharsets.UTF_8),
                    ),
                )

                else -> error("契約表が未対応の method を使っている: $method")
            }
            // 条件付き更新（段階5-4）。表に書けるのは固定文字列だけなので、
            // 「一致する etag を送る」側は BackendBehaviourTest が持つ。
            /* Content-Type（段階11-5）。ブラウザが送るヘッダで 415 にならないことを固定する */
            if (request.has("contentType")) {
                builder.header("Content-Type", request.path("contentType").asString())
            }
            if (request.has("ifMatch")) {
                builder.header("If-Match", request.path("ifMatch").asString())
            }
            if (request.has("ifNoneMatch")) {
                builder.header("If-None-Match", request.path("ifNoneMatch").asString())
            }
            return http.send(builder.build(), HttpResponse.BodyHandlers.ofByteArray())
        }
    }
}
